### PR TITLE
Cache lock

### PR DESCRIPTION
**Pull request by tommydcjung**

### CODE REVIEW

File: bsg_cache/bsg_cache_miss.v [hunk 1]
```diff
@@ -143,9 +146,16 @@ module bsg_cache_miss
           ? dirty_r
           : dirty_i;
         dma_send_fill_addr_o = 1'b1;
-        chosen_set_n = valid_v_i[0]
-          ? (valid_v_i[1] ? ~mru_n : 1'b1)
-          : 1'b0;
+        //chosen_set_n = valid_v_i[0]
+        //  ? (valid_v_i[1] ? ~mru_n : 1'b1)
+        //  : 1'b0;
+        chosen_set_n = lock_v_i[0]
```
Comment (taylor-bsg) [inline bsg_cache/bsg_cache_miss.v]: // maybe a small epsilon more clear; eliding the 1's and 0's
chosen_set_n = lock_v_i[0] // chose set 1 if set 0 is locked
                           | (~lock_v_i[1] // (otherwise chose set 0 if set 1 is locked)
                              & valid_v_i[0] // (otherwise chose set 0 if it is invalid)
                              & (~valid_v_i[1] | ~mru_n)) // otherwise chose set 1 if it is invalid, or if it is LRU; otherwise chose 0

// lock_v_i[0]  lock_v_i[1]  valid_v_i[0]  valid_v_i[1]  mru_n  set  (output) 
//    1                  X                X                  X               X          1
//   0                  1                 X                  X               X          0
//   0                  0                0                   X               X         0
//   0                  0                1                    0              X          1
//   0                  0                1                    1              0           1
//   0                  0                1                    1              1            0

File: bsg_mem/bsg_mem_1rw_sync_mask_write_byte.v [hunk 2]
```diff
@@ -1,63 +1,74 @@
-module bsg_mem_1rw_sync_mask_write_byte #( parameter els_p = -1
-                                          ,parameter addr_width_lp = `BSG_SAFE_CLOG2(els_p)
+/**
```
Comment (taylor-bsg) [inline bsg_mem/bsg_mem_1rw_sync_mask_write_byte.v]: Unnecessary reformat of the whole file. We prefer not to mangle the git history this if the code is more or less compliant..

File: bsg_mem/bsg_mem_1rw_sync_mask_write_byte_synth.v [hunk 3]
```diff
@@ -1,42 +1,47 @@
 // NOTE: Users of BaseJump STL should not instantiate this module directly
```
Comment (taylor-bsg) [inline bsg_mem/bsg_mem_1rw_sync_mask_write_byte_synth.v]: Too much reformatting; original code is basically compliant. Prefer not to have so much git churn.

FYI localparam in parameter list does not work on iverilog 0.10.0 or Cadence Incisive 15.20, which are relatively recent.

Comment (taylor-bsg) [reply bsg_mem/bsg_mem_1rw_sync_mask_write_byte_synth.v]: So we should avoid using it.

File: bsg_mem/bsg_mem_1rw_sync_mask_write_byte_synth.v [hunk 4]
```diff
@@ -1,42 +1,47 @@
...
 // they should use bsg_mem_1r1w_sync_mask_write_byte.
```
Comment (taylor-bsg) [inline bsg_mem/bsg_mem_1rw_sync_mask_write_byte_synth.v]: churn

File: bsg_mem/bsg_mem_1rw_sync_synth.v [hunk 5]
```diff
@@ -6,45 +6,56 @@
 // NOTE: Users of BaseJump STL should not instantiate this module directly
```
Comment (taylor-bsg) [inline bsg_mem/bsg_mem_1rw_sync_synth.v]: Avoid churn.

File: bsg_mem/bsg_mem_1rw_sync_synth.v [hunk 6]
```diff
@@ -6,45 +6,56 @@
...
-       addr_r <= addr_i;
-     else
-       addr_r <= 'X;
+  logic [width_p-1:0] mem [els_p-1:0];
+
+  always_ff @ (posedge clk_i) begin
+    if (v_i & ~w_i) begin
+      data_o <= mem[addr_i];
+    end
+    else begin
```
Comment (taylor-bsg) [inline bsg_mem/bsg_mem_1rw_sync_synth.v]: begin and end should align

File: bsg_mem/bsg_mem_1rw_sync_synth.v [hunk 7]
```diff
@@ -6,45 +6,56 @@
...
+      if (latch_last_read_p == 0) begin
+        data_o <= 'X;
+      end
+    end
+  end
 
-   assign data_o = mem[addr_r];
 
    always_ff @(posedge clk_i)
```
Comment (taylor-bsg) [inline bsg_mem/bsg_mem_1rw_sync_synth.v]: Should put the two blocking accesses to mem in the same always block.

The commits for cache locking and the modifications to ram semantics should be done in a separate pull request, with the cache locking one first. Before accepting the pull request for the rams, Ling/Scott should do a PPA regression to confirm that this is not adversely impacting the QOR of our register arrays.

File: bsg_dmc/bsg_dmc_phy.v [hunk 8]
```diff
@@ -453,8 +453,8 @@ module bsg_dmc_phy #
   assign ck_p_o     = clk_1x_0;
```
Comment (taylor-bsg) [inline bsg_dmc/bsg_dmc_phy.v]: DMC commits should be in separate pull request.

File: bsg_cache/bsg_cache.v [hunk 9]
```diff
@@ -11,10 +11,8 @@
  * See https://docs.google.com/document/d/15XJv3vH-08uvdUijjlAHJKznX4Wf42e19rZMyExlxGI/view for design doc
```
Comment (taylor-bsg) [inline bsg_cache/bsg_cache.v]: Nice mods! See comments for upgrades.

File: bsg_cache/bsg_cache_dma.v [hunk 10]
```diff
@@ -95,34 +94,11 @@ module bsg_cache_dma
     ,.yumi_i(in_fifo_yumi_li)
   );
```
Comment (taylor-bsg) [inline bsg_cache/bsg_cache_dma.v]: why did the FIFO go away?

File: bsg_cache/bsg_cache_dma.v [hunk 11]
```diff
@@ -212,32 +181,17 @@ module bsg_cache_dma
       end
 
       SEND_EVICT_DATA: begin
-        // counter_r in this context means the number of words read from
```
Comment (taylor-bsg) [inline bsg_cache/bsg_cache_dma.v]: why did this code go away?

File: bsg_mem/bsg_mem_1rw_sync.v [hunk 12]
```diff
@@ -4,47 +4,54 @@
 // Only one read or one write may be done per cycle.
```
Comment (taylor-bsg) [inline bsg_mem/bsg_mem_1rw_sync.v]: Unnecessary git churn?

File: bsg_mem/bsg_mem_1rw_sync_mask_write_bit.v [hunk 13]
```diff
@@ -3,50 +3,60 @@
 // Synchronous 1-port ram.
 // Only one read or one write may be done per cycle.
 
-module bsg_mem_1rw_sync_mask_write_bit #(parameter width_p=-1
-			               , parameter els_p=-1
-                     , parameter enable_clock_gating_p=0
-			               , parameter addr_width_lp=`BSG_SAFE_CLOG2(els_p)
-                     )
-   (input   clk_i
+module bsg_mem_1rw_sync_mask_write_bit
```
Comment (taylor-bsg) [inline bsg_mem/bsg_mem_1rw_sync_mask_write_bit.v]: Unnecessary git churn?

File: bsg_mem/bsg_mem_1rw_sync_mask_write_bit_synth.v [hunk 14]
```diff
@@ -7,31 +7,40 @@
...
-   always_ff @(posedge clk_i)
-     if (v_i & ~w_i)
-       addr_r <= addr_i;
-     else
-       addr_r <= 'X;
+  logic [width_p-1:0] mem [els_p-1:0];
+
+  always_ff @ (posedge clk_i) begin
+    if (v_i & ~w_i) begin
+      data_o <= mem[addr_i];
```
Comment (taylor-bsg) [inline bsg_mem/bsg_mem_1rw_sync_mask_write_bit_synth.v]: A little weird to put two accesses two mem in separate always_ff block, since there could be race. See also comments about putting these ram accesses in a separate commit, and testing PPA impact.

File: bsg_cache/bsg_cache.v [hunk 15]
```diff
@@ -190,17 +189,20 @@ module bsg_cache
 
 
   // tag_mem
+  // tag entry format = {valid, lock, tag}
```
[the remark this answers is not in the document]
Comment (tommydcjung) [reply bsg_cache/bsg_cache.v]: sure, but trying to be agile :)

File: bsg_mem/bsg_mem_1rw_sync_synth.v (hunk 7 above)
Comment (tommydcjung) [reply bsg_mem/bsg_mem_1rw_sync_synth.v]: I'm not sure which two blocking access you are referring to. are you talking about "data_o <= mem[addr_i]" and "mem[addr_i] <= data_i"?

File: bsg_cache/bsg_cache_dma.v (hunk 10 above)
Comment (tommydcjung) [reply bsg_cache/bsg_cache_dma.v]: added it back

File: bsg_cache/bsg_cache_dma.v (hunk 11 above)
Comment (tommydcjung) [reply bsg_cache/bsg_cache_dma.v]: added it back with fifo
